### PR TITLE
command buckets converted to the cobra style

### PR DESCRIPTION
**Pull request by tommyshem**
issue #472 Migrate all commands to cobra style commands

converted buckets command to cobra style

### CODE REVIEW

**Comment by ivanvc** [comment]
Hi @tommyshem, thanks for your pull request. Could you please sign your commit? So the developer certificate of origin (DCO) check passes, i.e:

```
git rebase HEAD~1 --signoff
git push --force
```

Thanks!

Ref: https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#commit-your-change

/ok-to-test

**Comment by ahrtr** [comment]
Thanks for raising the PR.  It doesn't break the user experience, so I agree to get this PR included in 1.4. 

Please resolve the test failure in case `TestBucketsCommand_Run`.

File: cmd/bbolt/command_version.go
Comment (tommyshem) [inline cmd/bbolt/command_version.go:1]: changed the output so it could be tested in the future

File: cmd/bbolt/main_test.go
Comment (tommyshem) [inline cmd/bbolt/main_test.go:1]: removed the tests and put them in the buckets test file

**Comment by ahrtr** [comment]
Please signoff the commit, read https://github.com/etcd-io/bbolt/pull/859/checks?check_run_id=32833799850

**Comment by tommyshem** [comment]
Should be finished now for review.
But some how I upload the bbolt.exe in the pull request and I don't know how to remove it.

File: cmd/bbolt/bbolt.exe
[the remark this answers is not in the document]
Comment (ivanvc) [reply cmd/bbolt/bbolt.exe:1]: I think running `git rm cmd/bbolt/bbolt.exe` should work, as the file is not ignored. If it doesn't work, you can try adding `-f`, i.e., `git rm -f cmd/bbolt/bbolt.exe`.

You'll then need to do a commit.

File: cmd/bbolt/command_version.go [hunk 1]
```diff
@@ -15,9 +15,9 @@ func newVersionCommand() *cobra.Command {
 		Short: "print the current version of bbolt",
 		Long:  "print the current version of bbolt",
 		Run: func(cmd *cobra.Command, args []string) {
-			fmt.Printf("bbolt Version: %s\n", version.Version)
-			fmt.Printf("Go Version: %s\n", runtime.Version())
-			fmt.Printf("Go OS/Arch: %s/%s\n", runtime.GOOS, runtime.GOARCH)
+			fmt.Fprintln(cmd.OutOrStdout(), "bbolt Version: ", version.Version)
+			fmt.Fprintln(cmd.OutOrStdout(), "Go Version: ", runtime.Version())
+			fmt.Fprintln(cmd.OutOrStdout(), "Go OS/Arch: ", runtime.GOOS, "/", runtime.GOARCH)
```
Comment (ahrtr) [inline cmd/bbolt/command_version.go:20]: One PR only does one thing. Please don't make unrelated change.

**Comment by tommyshem** [comment]
Removed the bolt.exe file and the command_version.go file.
So it should be  just the buckets command change now hopefully.

File: cmd/bbolt/command_buckets.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	bolt "go.etcd.io/bbolt"
+)
+
+// newBucketsCommand creates a new command that prints a list of buckets in the given Bolt database.
+//
+// The path to a Bolt database must be specified as an argument.
```
Comment (ahrtr) [inline cmd/bbolt/command_buckets.go:13]: It's unnecessary comment to me. The command is self-explanation.
```suggestion
```

File: cmd/bbolt/command_buckets.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+func newBucketsCommand() *cobra.Command {
+	var bucketsCmd = &cobra.Command{
+		Use:   "buckets <path>",
+		Short: "Print a list of buckets",
+		Long:  "Print a list of buckets in the given Bolt database\nThe path to a Bolt database must be specified as an argument",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return printBucketsList(cmd, args[0])
+		},
```
Comment (ahrtr) [inline cmd/bbolt/command_buckets.go:22]: Please try to keep consistent with other commands implementation.
```suggestion
		Short: "print a list of buckets",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return bucketsFunc(args[0])
		},
```

File: cmd/bbolt/command_buckets.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+// printBucketsList prints a list of buckets in the given Bolt database.
+func printBucketsList(cmd *cobra.Command, path string) error {
+	// Required database path.
+	if path == "" {
+		return ErrPathRequired
+		// Verify if the specified database file exists.
+	} else if _, err := os.Stat(path); os.IsNotExist(err) {
+		return ErrFileNotFound
+	}
```
Comment (ahrtr) [inline cmd/bbolt/command_buckets.go:35]: Please use `checkSourceDBPath`. See example below,
https://github.com/etcd-io/bbolt/blob/cb0618b9da2b436b7bed6013b515697cbaeb458a/cmd/bbolt/command_inspect.go#L27

File: cmd/bbolt/command_buckets_test.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	rootCmd := main.NewRootCommand()
+	_, actualOutput, err := executeCommand(rootCmd, "buckets", db.Path()) //rootCmd, "buckets", db.Path())
+	require.NoError(t, err)
+	t.Log("Verify result")
+	expected := "bar\nbaz\nfoo\n"
+	require.EqualValues(t, expected, actualOutput)
+}
+
+// executeCommand runs the given command and returns the output and error.
+func executeCommand(rootCmd *cobra.Command, args ...string) (*cobra.Command, string, error) {
```
Comment (ahrtr) [inline cmd/bbolt/command_buckets_test.go:46]: Can we just remove this function and get the implementation inlined in the test?

If you want to make it as a generic test utility function (i.e. included in utils_test.go), let's do it in a followup PR and we can discuss it separately.

Comment (tommyshem) [reply cmd/bbolt/command_buckets_test.go:46]: I have inline it as requested.

File: cmd/bbolt/command_buckets.go (hunk 4 above)
Comment (tommyshem) [reply cmd/bbolt/command_buckets.go:31]: I used the original code but I can look at changing it to the above

**Comment by tommyshem** [comment]
All the changes from above should be done now.

**Comment by ahrtr** [comment]
@tommyshem you need to rebase this PR instead of merging main branch into your dev branch. Please also squash the commits into one.

**Comment by tommyshem** [comment]
Sorry new to github pull requests as only used github mainly for personal projects. I will try tomorrow to rebase and squash commits. Do I need to put any info in the commit text or just reference the issue number.

File: cmd/bbolt/command_buckets.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+			return bucketsFunc(cmd, args[0])
+		},
+	}
+	return bucketsCmd
+}
+
+func bucketsFunc(cmd *cobra.Command, srcDBPath string) error {
+	// Required database path.
+	if srcDBPath == "" {
+		return ErrPathRequired
```
Comment (ahrtr) [inline cmd/bbolt/command_buckets.go:25]: I think the `cobra.ExactArgs(1)` already guarantees that it won't be empty string? Pleasae follow the same pattern as other command.

File: cmd/bbolt/command_buckets.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+	db, err := bolt.Open(srcDBPath, 0600, &bolt.Options{ReadOnly: true})
+	if err != nil {
+		return err
+	}
+	defer db.Close()
+
+	// Print the list of buckets in the database.
+	return db.View(func(tx *bolt.Tx) error {
+		return tx.ForEach(func(name []byte, _ *bolt.Bucket) error {
+			fmt.Fprintln(cmd.OutOrStdout(), string(name))
```
Comment (ahrtr) [inline cmd/bbolt/command_buckets.go:41]: I agree that `cmd.OutOrStdout()` is a little better than `os.Stdout`, but please keep it consistent with other commands. We can replace them with `cmd.OutOrStdout()` in separate PR.

File: cmd/bbolt/command_buckets.go (hunk 6 above)
Comment (tommyshem) [reply cmd/bbolt/command_buckets.go:25]: This check is from the original code.

File: cmd/bbolt/command_buckets.go (hunk 7 above)
Comment (tommyshem) [reply cmd/bbolt/command_buckets.go:41]: cmd.OutOrStdout() is all ready used in the cobra style converted files already
command_check.go line 44, 65, 70
command_get.go line 46, 92
command_info.go line 47
command_version.go line 18, 19, 20

File: cmd/bbolt/command_buckets.go (hunk 6 above)
Comment (ahrtr) [reply cmd/bbolt/command_buckets.go:25]: "the original code" doesn't mean that it's necessary.

Please also always remember to simplify the the implementation,

https://github.com/etcd-io/bbolt/blob/cb0618b9da2b436b7bed6013b515697cbaeb458a/cmd/bbolt/command_inspect.go#L26-L30

File: cmd/bbolt/command_buckets.go (hunk 7 above)
Comment (ahrtr) [reply cmd/bbolt/command_buckets.go:41]: I see that it's only used in `command_check.go`. I should have raised this comment in [PR 723](https://github.com/etcd-io/bbolt/pull/723).

Overall, not a big deal. Feel free to keep it as it's

Comment (tommyshem) [reply cmd/bbolt/command_buckets.go:38]: I must have changed the over files in my copy above when I was testing the outputs.
As I had trouble capturing the output when I had converted to the cobra style without the cmd.OutOrStdout() function

**Comment by ahrtr** [comment]
This PR is almost ready to be approved & merged. 

Please squash the commits and ensure it's signed off. thx
